### PR TITLE
RedisPipeline ignore flush when call dtor from another thread.

### PR DESCRIPTION
**Pull request by liuh-80**
#### Why I did it
Because RedisPipeline is not thread safe, in c++ if code call exit() from another thread, the dtor of RedisPipeline will  be called from another thread and trigger a race condition issue.

#### How I did it
RedisPipeline ignore flush when call dtor from another thread.

#### How to verify it
Pass all UT and E2E test cases.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

#### Description for the changelog
RedisPipeline ignore flush when call dtor from another thread.

### CODE REVIEW

File: common/redispipeline.h [hunk 1]
```diff
@@ -7,6 +7,10 @@
 #include "rediscommand.h"
 #include "dbconnector.h"
 
+#include "unistd.h"
+#include "sys/syscall.h"
+#define gettid() syscall(SYS_gettid)
```
Comment (liuh-80) [inline common/redispipeline.h:12]: there is no wrapper for gettid() in glibc <= 2.30, so we need use syscall(SYS_gettid)

File: common/redispipeline.h [hunk 2]
```diff
@@ -19,10 +23,16 @@ class RedisPipeline {
...
+        initializeOwnerTid();
     }
 
     ~RedisPipeline() {
-        flush();
+        if (m_ownerTid == gettid())
+        {
+            // call flush from different thread will trigger race condition issue.
+            flush();
+        }
```
Comment (qiluo-msft) [inline common/redispipeline.h:34]: >} [](http://example.com/codeflow?start=8&length=1)

Add a notice logging in `else` branch.

File: common/redispipeline.h [hunk 3]
```diff
@@ -19,10 +23,16 @@ class RedisPipeline {
         , m_remaining(0)
     {
         m_db = db->newConnector(NEWCONNECTOR_TIMEOUT);
+        initializeOwnerTid();
     }
 
     ~RedisPipeline() {
```
Comment (qiluo-msft) [inline common/redispipeline.h:29]: >RedisPipeline [](http://example.com/codeflow?start=5&length=13)

Thanks for fixing the bug. It is from the design that DBConnector could be used only in one thread. Do you find similar issues in other classes in swss-common?

Comment (liuh-80) [reply common/redispipeline.h:29]: No, I check all dtor in swss-common only find RedisPipeline class run redis command in dtor.

File: common/redispipeline.h (hunk 2 above)
Comment (liuh-80) [reply common/redispipeline.h:35]: Fixed

**Review by qiluo-msft — APPROVED** [review]

**Comment by qiluo-msft** [comment]
Do we need backport, for example to 202012 branch?

**Pull request merged**